edalize/libero.py

```diff
@@ -0,0 +1,181 @@
+import logging
+import os
+from edalize.edatool import Edatool
+
+logger = logging.getLogger(__name__)
+
+
+class Libero(Edatool):
+    @classmethod
+    def get_doc(cls, api_ver):
+        if api_ver == 0:
+            return {'description': "The Libero backend supports Microsemi Libero to build systems and program the FPGA",
+                    'members': [
+                        {'name': 'family',
+                         'type': 'String',
+                         'desc': 'FPGA family (e.g. PolarFire)'},
+                        {'name': 'die',
+                         'type': 'String',
+                         'desc': 'FPGA device (e.g. MPF300TS)'},
+                        {'name': 'package',
+                         'type': 'String',
+                         'desc': 'FPGA package type (e.g. FCG1152)'},
+                        {'name': 'speed',
+                         'type': 'String',
+                         'desc': 'FPGA speed rating (e.g. -1)'},
+                        {'name': 'dievoltage',
+                         'type': 'String',
+                         'desc': 'FPGA die voltage (e.g. 1.0)'},
+                        {'name': 'range',
+                         'type': 'String',
+                         'desc': 'FPGA temperature range (e.g. IND)'},
+                        {'name': 'defiostd',
+                         'type': 'String',
+                         'desc': 'FPGA default IO std (e.g. "LVCMOS 1.8V"'},
+                        {'name': 'hdl',
+                         'type': 'String',
+                         'desc': 'Default HDL (e.g. "VERILOG"'},
+                        {'name': 'synth_options',
+                         'type': 'String',
+                         'desc': 'Additional Synthesize tool options separated by commas (e.g. "RETIMING:true,CLOCK_GLOBAL:2")'},
+                        {'name': 'pnr_options',
```

Comment:
While this helps for setting pnr options, there are just too many things that different users want to change and we can't expose all these things. What we do instead is to make it possible for users to add any extra options they want to set as tcl files that we source from the project creation script. We do this by looking for files with `file_type : tclSource` and source them. You can compare with e.g. the quartus, ise or vivado backends which offer the same possibilities.

The situation is a bit more complicated with the synthesis options as we need to mix some hard-coded commands together with user-supplied ones. I think I have an idea but it's not super pretty

Reply:
Done, added sourcing user TCL scripts. Gonna remove the synth and pnr opts.

Reply:
Done! TCL scripts are sourced from main script.